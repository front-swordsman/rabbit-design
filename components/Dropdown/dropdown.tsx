import * as React from 'react'
import { DropdownProps } from 'types/dropdown'
import Portal from '../Portal'

import './index.scss'

const defaultProps: Partial<DropdownProps> = {
  disabled: false,
  visible: false,
  trigger: 'hover',
  placement: 'bottomLeft',
  destroy: true
}

const Dropdown: React.FC<DropdownProps> = dropdown => {
  const props = { ...defaultProps, ...dropdown }
  return <Portal {...props} prefixCls="snake-dropdown" mode="dropdown" />
}

export default Dropdown
