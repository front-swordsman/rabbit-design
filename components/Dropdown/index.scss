@import '../styles/themes/default';

$dropdown: 'snake-dropdown';

.#{$dropdown} {
  position: absolute;
  z-index: $zindexDropdown;
  box-sizing: border-box;
  color: rgba(0, 0, 0, 0.65);
  font-size: $body;
  line-height: $lineHeight;
  list-style: none;
  cursor: pointer;
  background-color: #fff;
  margin: 0;
  padding: 4px 0;
  text-align: left;

  &-content {
    box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15);
    border-radius: $borderRadius;
    padding: 4px 0;
  }

  &-slide-appear,
  &-slide-enter {
    opacity: 0;
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: $easeOutQuint;
    animation-play-state: paused;
  }

  &-slide-appear-active.#{$dropdown}-bottom,
  &-slide-appear-active.#{$dropdown}-bottomLeft,
  &-slide-appear-active.#{$dropdown}-bottomRight,
  &-slide-enter-active.#{$dropdown}-bottom,
  &-slide-enter-active.#{$dropdown}-bottomLeft,
  &-slide-enter-active.#{$dropdown}-bottomRight {
    animation-name: slideUpIn;
    animation-play-state: running;
  }

  &-slide-appear-active.#{$dropdown}-top,
  &-slide-appear-active.#{$dropdown}-topLeft,
  &-slide-appear-active.#{$dropdown}-topRight,
  &-slide-enter-active.#{$dropdown}-top,
  &-slide-enter-active.#{$dropdown}-topLeft,
  &-slide-enter-active.#{$dropdown}-topRight {
    animation-name: slideDownIn;
    animation-play-state: running;
  }

  &-slide-exit {
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: $easeInQuint;
    animation-play-state: paused;
  }

  &-slide-exit-active.#{$dropdown}-bottom,
  &-slide-exit-active.#{$dropdown}-bottomLeft,
  &-slide-exit-active.#{$dropdown}-bottomRight {
    animation-name: slideUpOut;
    animation-play-state: running;
  }

  &-slide-exit-active.#{$dropdown}-top,
  &-slide-exit-active.#{$dropdown}-topLeft,
  &-slide-exit-active.#{$dropdown}-topRight {
    animation-name: slideDownOut;
    animation-play-state: running;
  }
}

@keyframes slideUpIn {
  0% {
    opacity: 0;
    transform: scaleY(0.8);
    transform-origin: 0% 0%;
  }

  100% {
    opacity: 1;
    transform: scaleY(1);
    transform-origin: 0% 0%;
  }
}

@keyframes slideUpOut {
  0% {
    opacity: 1;
    transform: scaleY(1);
    transform-origin: 0% 0%;
  }

  100% {
    opacity: 0;
    transform: scaleY(0.8);
    transform-origin: 0% 0%;
  }
}

@keyframes slideDownIn {
  0% {
    opacity: 0;
    transform: scaleY(0.8);
    transform-origin: 100% 100%;
  }

  100% {
    opacity: 1;
    transform: scaleY(1);
    transform-origin: 100% 100%;
  }
}

@keyframes slideDownOut {
  0% {
    opacity: 1;
    transform: scaleY(1);
    transform-origin: 100% 100%;
  }

  100% {
    opacity: 0;
    transform: scaleY(0.8);
    transform-origin: 100% 100%;
  }
}
