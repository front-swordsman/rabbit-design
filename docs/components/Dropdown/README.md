---
imports: 
  import DropdownDemoCode from '!raw-loader!./demo.tsx';
  import DropdownDemo from './demo.tsx';
  import DropdownBasicCode from '!raw-loader!./basic.tsx';
  import DropdownBasic from './basic.tsx';
---

# Dropdown

## 基础用法

:::demo
<Block code={DropdownBasicCode} des="基础用法">
<DropdownBasic />
</Block>
:::

## 简单示例

:::demo
<Block code={DropdownDemoCode} des="复选框">
<DropdownDemo />
</Block>
:::

## Props

| 参数             | 说明                             | 类型                                                                | 可选值 | 默认值       |
| ---------------- | -------------------------------- | ------------------------------------------------------------------- | ------ | ------------ |
| disabled         | 是否禁用                         | boolean                                                             | --     | false        |
| visible          | 是否可见                         | boolean                                                             | --     | false        |
| content          | 下拉内容区                       | ReactNode                                                           | --     | --           |
| wrapperComponent | 触发元素的包裹层， 默认为 `span` | React.ElementType                                                   | --     | `span`       |
| wrapperStyle     | 触发元素的包裹层的样式           | object                                                              | --     | --           |
| trigger          | 触发弹出层出现的方式             | `hover`, `click`                                                    | --     | `hover`      |
| placement        | 弹出层位置                       | `top`, `bottom`, `topLeft`, `topRight`, `bottomLeft`, `bottomRight` | --     | `bottomLeft` |
| onVisibleChange  | 显影发生改变的回调               | (visible: boolean) => void                                          | --     | --           |
| destroy          | 弹层小时是否销毁                 | boolean                                                             | --     | `true`       |
| className        | 弹出层的 `class`                 | string                                                              | --     | --           |
| style            | 弹出层的样式                     | object                                                              | --     | --           |