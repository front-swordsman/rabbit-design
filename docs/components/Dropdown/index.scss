.drop-demo {
  height: 200px;
  overflow-y: scroll;

  &-content {
    height: 600px;
    background-image: url(https://zos.alipayobjects.com/rmsportal/RmjwQiJorKyobvI.jpg);
    padding-top: 100px;
  }
}