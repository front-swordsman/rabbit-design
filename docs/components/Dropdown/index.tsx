import * as React from 'react'
import DropdownDoc from './README.md'

export default function IconView() {
  return <DropdownDoc />
}
