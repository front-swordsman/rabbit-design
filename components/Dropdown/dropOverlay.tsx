import * as React from 'react'
import * as ReactDOM from 'react-dom'
import { CSSTransition } from 'react-transition-group'
import cx from 'classnames'
import { DropDownOverlay } from 'types/dropdown'
import { useClickOutSide } from '../utils/use'

import './index.scss'

const defaultProps = {
  trigger: 'hover',
  placement: 'bottomLeft',
  disabled: false,
  wrapperComponent: 'span',
  prefixCls: 'snake-dropdown',
  hasTriangle: false,
  timeout: 300,
  offset: 0,
  mode: 'dropdown'
}

const DropOverlay: React.FC<DropDownOverlay> = dropdownProps => {
  const props = { ...defaultProps, ...dropdownProps }
  const {
    className,
    placement,
    content,
    wrapperComponent: Component,
    children,
    trigger,
    disabled,
    getPopupContainer,
    prefixCls,
    animationName,
    wrapperStyle,
    // style,
    onVisibleChange,
    hasTriangle,
    timeout,
    offset,
    mode
  } = props
  const [visible, setVisible] = React.useState(false)
  const wrapperRef = React.useRef<HTMLElement>()
  const contentRef = React.useRef<HTMLDivElement>()

  useClickOutSide([contentRef, wrapperRef], () => {
    setVisible(false)
  })

  const toggle = (triggerStr: string, toggle: boolean) => {
    if (disabled) return
    if (trigger === triggerStr) {
      setVisible(toggle)
      // 只在变化的时候回调
      if (toggle !== visible) {
        onVisibleChange && onVisibleChange(toggle)
      }
    }
  }

  const getContainer = () => {
    if (getPopupContainer) {
      return getPopupContainer()
    }
    return document.body
  }

  const getOverlayClass = React.useCallback(() => {
    if (animationName) return `${prefixCls}-${animationName}`
    return `${prefixCls}-slide`
  }, [animationName])

  const getContentStyle = (top: string, left: string) => {
    ;(contentRef.current as HTMLElement).style.top = top
    ;(contentRef.current as HTMLElement).style.left = left
  }

  const getPosition = () => {
    const rect = (wrapperRef.current as HTMLElement).getBoundingClientRect()
    const contentRect = (contentRef.current as HTMLElement).getBoundingClientRect()
    let top = `${rect.top + rect.height + offset + window.pageYOffset}px`
    let left = `${rect.left + window.pageXOffset}px`
    if (contentRect.width < rect.width && mode === 'dropdown') {
      contentRef.current.style.minWidth = `${rect.width}px`
    }
    switch (placement) {
      case 'top':
        top = `${rect.top - offset - contentRect.height + window.pageYOffset}px`
        left = `${rect.left + rect.width / 2 - contentRect.width / 2 + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'topLeft':
        top = `${rect.top - offset - contentRect.height + window.pageYOffset}px`
        left = `${rect.left + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'topRight':
        top = `${rect.top - offset - contentRect.height + window.pageYOffset}px`
        left = `${rect.right - contentRect.width + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'bottom':
        top = `${rect.top + rect.height + offset + window.pageYOffset}px`
        left = `${rect.left + rect.width / 2 - contentRect.width / 2 + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'bottomLeft':
        top = `${rect.top + rect.height + offset + window.pageYOffset}px`
        left = `${rect.left + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'bottomRight':
        top = `${rect.top + rect.height + offset + window.pageYOffset}px`
        left = `${rect.right - contentRect.width + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'left':
        top = `${rect.top + rect.height / 2 - contentRect.height / 2 + window.pageYOffset}px`
        left = `${rect.left - offset - contentRect.width + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'leftTop':
        top = `${rect.top + window.pageYOffset}px`
        left = `${rect.left - offset - contentRect.width + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'leftBottom':
        top = `${rect.top + rect.height - contentRect.height + window.pageYOffset}px`
        left = `${rect.left - offset - contentRect.width + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'right':
        top = `${rect.top + rect.height / 2 - contentRect.height / 2 + window.pageYOffset}px`
        left = `${rect.right + offset + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'rightTop':
        top = `${rect.top + window.pageYOffset}px`
        left = `${rect.right + offset + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      case 'rightBottom':
        top = `${rect.top + rect.height - contentRect.height + window.pageYOffset}px`
        left = `${rect.right + offset + window.pageXOffset}px`
        getContentStyle(top, left)
        break
      default:
        getContentStyle(top, left)
        break
    }
  }

  const handleClick = () => {
    toggle('click', true)
  }

  const handleMouseEnter = () => {
    toggle('hover', true)
  }

  const handleMouseLeave = () => {
    toggle('hover', false)
  }

  const renderDropDown = () => {
    const classStr = cx(
      prefixCls,
      {
        [`${prefixCls}-${placement}`]: placement,
        [`${prefixCls}-triangle`]: hasTriangle
      },
      className
    )
    return (
      <CSSTransition in={visible} timeout={timeout} classNames={getOverlayClass()} unmountOnExit>
        <div
          className={classStr}
          ref={contentRef}
          onClick={() => {
            setVisible(false)
            onVisibleChange && onVisibleChange(false)
          }}
          onMouseEnter={() => toggle('hover', true)}
          onMouseLeave={() => toggle('hover', false)}
        >
          {content}
        </div>
      </CSSTransition>
    )
  }

  React.useLayoutEffect(() => {
    if (visible) {
      getPosition()
    }
  }, [visible])

  return (
    <>
      <Component
        onClick={handleClick}
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        style={wrapperStyle}
        ref={wrapperRef}
      >
        {React.cloneElement(children as any, {
          disabled: disabled
        })}
      </Component>
      {ReactDOM.createPortal(renderDropDown(), getContainer())}
    </>
  )
}

export default DropOverlay
