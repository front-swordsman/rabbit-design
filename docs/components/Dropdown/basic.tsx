import * as React from 'react'
import { Dropdown } from 'components'

export default function BasicDemo() {
  const [visible, setVisible] = React.useState(false)
  return (
    <div>
      <Dropdown
        visible={visible}
        onVisibleChange={visible => setVisible(visible)}
        content={
          <div>
            <div onClick={() => console.log('click')}>这是内容区</div>
            <div>这是内容区</div>
            <div>这是内容区</div>
          </div>
        }
      >
        <a href="">hover me</a>
      </Dropdown>
    </div>
  )
}
