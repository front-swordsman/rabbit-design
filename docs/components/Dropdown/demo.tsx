import * as React from 'react'
import { Button, Radio, Dropdown } from 'components'

import './index.scss'

export default function DropdownDemo() {
  const [action, setAction] = React.useState<any>('click')
  const [visible, setVisible] = React.useState(false)
  const [direction, setDirection] = React.useState<any>('bottomLeft')
  return (
    <div>
      <div>
        <Radio
          options={[{ label: 'click', value: 'click' }, { label: 'hover', value: 'hover' }]}
          value={action}
          onChange={(value: string) => setAction(value)}
        />
      </div>
      <div>
        <Radio
          options={[
            { label: 'bottomLeft', value: 'bottomLeft' },
            { label: 'bottom', value: 'bottom' },
            { label: 'bottomRight', value: 'bottomRight' },
            { label: 'topLeft', value: 'topLeft' },
            { label: 'top', value: 'top' },
            { label: 'topRight', value: 'topRight' }
          ]}
          value={direction}
          onChange={(value: string) => setDirection(value)}
        />
      </div>
      <div style={{ marginTop: 20 }}>
        <Dropdown
          content={
            <div>
              <div>数字数字数字数字数字</div>
              <div>数字数字数字数字数字</div>
              <div>数字数字数字数字数字</div>
            </div>
          }
          visible={visible}
          trigger={action}
          placement={direction}
          wrapperComponent="div"
          wrapperStyle={{ display: 'inline-block' }}
          onVisibleChange={visible => {
            setVisible(visible)
          }}
        >
          <Button>{direction}</Button>
        </Dropdown>
      </div>
    </div>
  )
}
