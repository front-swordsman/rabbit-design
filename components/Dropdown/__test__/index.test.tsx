import * as React from 'react'
import { render } from 'react-testing-library'
import Dropdown from '../index'
import { Placement } from 'types/dropdown'

describe('Dropdown Test', () => {
  function DropdownDemo({
    placement = 'bottomLeft',
    visible = true
  }: {
    placement?: Placement
    visible?: boolean
  }) {
    return (
      <Dropdown placement={placement} visible={visible} content={<div>内容区</div>}>
        <span>click me</span>
      </Dropdown>
    )
  }

  it('snapshot when visible is true', () => {
    const { baseElement } = render(<DropdownDemo />)
    expect(baseElement).toMatchSnapshot()
  })

  it('snapshot when visible is false', () => {
    const { baseElement } = render(<DropdownDemo visible={false} />)
    expect(baseElement).toMatchSnapshot()
  })

  it('snapshot when placement is bottom', () => {
    const { baseElement } = render(<DropdownDemo placement="bottom" />)
    expect(baseElement).toMatchSnapshot()
  })

  it('snapshot when placement is bottomRight', () => {
    const { baseElement } = render(<DropdownDemo placement="bottomRight" />)
    expect(baseElement).toMatchSnapshot()
  })

  it('snapshot when placement is top', () => {
    const { baseElement } = render(<DropdownDemo placement="top" />)
    expect(baseElement).toMatchSnapshot()
  })

  it('snapshot when placement is topLeft', () => {
    const { baseElement } = render(<DropdownDemo placement="topLeft" />)
    expect(baseElement).toMatchSnapshot()
  })

  it('snapshot when placement is topRight', () => {
    const { baseElement } = render(<DropdownDemo placement="topRight" />)
    expect(baseElement).toMatchSnapshot()
  })
})
