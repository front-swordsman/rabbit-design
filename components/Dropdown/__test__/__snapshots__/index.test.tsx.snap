// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Test snapshot when placement is bottom 1`] = `
<body>
  <div>
    <span>
      <span>
        click me
      </span>
    </span>
  </div>
  <div
    class="snake-dropdown snake-dropdown-bottom"
    style="top: 0px; left: 0px;"
  >
    <div
      class="snake-dropdown-content"
    >
      <div>
        内容区
      </div>
    </div>
  </div>
</body>
`;

exports[`Dropdown Test snapshot when placement is bottomRight 1`] = `
<body>
  <div>
    <span>
      <span>
        click me
      </span>
    </span>
  </div>
  <div
    class="snake-dropdown snake-dropdown-bottomRight"
    style="top: 0px; left: 0px;"
  >
    <div
      class="snake-dropdown-content"
    >
      <div>
        内容区
      </div>
    </div>
  </div>
</body>
`;

exports[`Dropdown Test snapshot when placement is top 1`] = `
<body>
  <div>
    <span>
      <span>
        click me
      </span>
    </span>
  </div>
  <div
    class="snake-dropdown snake-dropdown-top"
    style="top: 0px; left: 0px;"
  >
    <div
      class="snake-dropdown-content"
    >
      <div>
        内容区
      </div>
    </div>
  </div>
</body>
`;

exports[`Dropdown Test snapshot when placement is topLeft 1`] = `
<body>
  <div>
    <span>
      <span>
        click me
      </span>
    </span>
  </div>
  <div
    class="snake-dropdown snake-dropdown-topLeft"
    style="top: 0px; left: 0px;"
  >
    <div
      class="snake-dropdown-content"
    >
      <div>
        内容区
      </div>
    </div>
  </div>
</body>
`;

exports[`Dropdown Test snapshot when placement is topRight 1`] = `
<body>
  <div>
    <span>
      <span>
        click me
      </span>
    </span>
  </div>
  <div
    class="snake-dropdown snake-dropdown-topRight"
    style="top: 0px; left: 0px;"
  >
    <div
      class="snake-dropdown-content"
    >
      <div>
        内容区
      </div>
    </div>
  </div>
</body>
`;

exports[`Dropdown Test snapshot when visible is false 1`] = `
<body>
  <div>
    <span>
      <span>
        click me
      </span>
    </span>
  </div>
</body>
`;

exports[`Dropdown Test snapshot when visible is true 1`] = `
<body>
  <div>
    <span>
      <span>
        click me
      </span>
    </span>
  </div>
  <div
    class="snake-dropdown snake-dropdown-bottomLeft"
    style="top: 0px; left: 0px;"
  >
    <div
      class="snake-dropdown-content"
    >
      <div>
        内容区
      </div>
    </div>
  </div>
</body>
`;
